### PR TITLE
RavenDB-20268: Prerequisites for implementation of streaming operations

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20268
https://issues.hibernatingrhinos.com/issue/RavenDB-20143

### Additional description

Improvements of resource consumption at the term level and implementation of reverse term lookup for ensuring `Fill` operations will respect specified order. This allow further implementation of streaming operations in the reverse order. 

### Type of change
- Optimization

### How risky is the change?
- Low

### CODE REVIEW

**Review by ayende — APPROVED** [review]

**Comment by arekpalinski** [comment]
@redknightlois This PR is ready to merge but in order to fix the issue https://issues.hibernatingrhinos.com/issue/RavenDB-20268 we need to incorporate the reverse iterator into the handling of a query at higher level. Is that correct?

**Comment by redknightlois** [comment]
Yes. This is just prerequisite work. I have other commits that are going to come on a different PR before @maciejaszyk can introduce the changes at the RavenDB query creation level.

**Comment by arekpalinski** [comment]
@redknightlois Please check failing tests:
https://github.com/ravendb/ravendb/actions/runs/4621105076/jobs/8172118602
https://github.com/ravendb/ravendb/actions/runs/4621105076/jobs/8172118783
https://github.com/ravendb/ravendb/actions/runs/4621105076/jobs/8172117865
https://github.com/ravendb/ravendb/actions/runs/4621105076/jobs/8172118054?pr=16177

**Comment by arekpalinski** [comment]
@redknightlois can you check failing tests and rebase the PR against v6.0 branch?

File: src/Corax/Queries/SpatialMatch/SpatialMatch.cs [hunk 1]
```diff
@@ -29,6 +29,9 @@ public class SpatialMatch : IQueryMatch
     private IDisposable _startsWithDisposeHandler;
     private HashSet<long> _alreadyReturned;
 
+    // TODO: Check if there is any way we can consider something ordered.
+    public bool IsOrdered => false;
```
Comment (maciejaszyk) [inline src/Corax/Queries/SpatialMatch/SpatialMatch.cs:33]: Geohash is using Z-order.

**Comment by maciejaszyk** [comment]
LGTM

**Review by maciejaszyk — APPROVED** [review]

**Comment by redknightlois** [comment]
Rebased on top of 6.0. Do I rebuild the PR or you can change it?

**Comment by redknightlois** [comment]
I will investigate the errors but defer the merge until @ayende changes at the compact trees from the Sorting PR. Those are extensive and I have my doubts if it makes sense to merge this commit until he is done.

**Comment by arekpalinski** [comment]
OK. Makes sense. Closing this PR for now

**Pull request closed without merging**